docs/zephyr_02_functional_requirements.sdoc

```diff
@@ -960,25 +960,184 @@ As a Zephyr OS user I want to be able to exchange 1 to N data objects between di
+shall provide a mechanism to lock a mutex.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-4
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Exclusive Locking of a Mutex
+STATEMENT: >>>
+When a mutex is successfully locked by a thread, the Zephyr RTOS shall ensure
+that the mutex becomes unavailable for locking by other threads until it is
+unlocked.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-5
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Unlocking of a Mutex
+STATEMENT: >>>
+While a thread no longer requires exclusive access to a shared resource, the
+Zephyr RTOS shall provide a mechanism to unlock a mutex.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-6
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Availability of a Mutex after Unlocking
+STATEMENT: >>>
+When a mutex is successfully unlocked by a thread, the Zephyr RTOS shall ensure
+that the mutex becomes available for locking by other threads.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-7
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Timed locking of a Mutex
+STATEMENT: >>>
+Mutexes shall support timed locking, where threads can specify a timeout period for waiting on the mutex.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-8
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Indifinite locking of a Mutex
+STATEMENT: >>>
+Mutexes shall support indefinite locking allowing threads to wait indefinitely until the mutex becomes available.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-9
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Reentrant locking
+STATEMENT: >>>
+The Zephyr RTOS shall support reentrant locking, allowing a thread to lock a mutex it has already locked without causing a deadlock.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-10
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Lock count
+STATEMENT: >>>
+Mutexes shall maintain a lock count to track the number of times a thread has locked the mutex.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-11
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Priority Inheritance
+STATEMENT: >>>
+When using mutexes for resource synchronization, the Zephyr RTOS shall implement priority inheritance protocols to prevent priority inversion scenarios.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-12
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Priority Inheritance - priority elevation
+STATEMENT: >>>
+When a higher-priority thread begins waiting on the mutex, the Zephyr RTOS shall temporarily elevate the priority of the owning thread.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-13
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Mutex Ownership
+STATEMENT: >>>
+Mutexes shall track the owning thread when locked to ensure exclusive access to the associated resource.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-MUTEX-14
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Mutex
+TITLE: Mutex Ownership - Unlock
+STATEMENT: >>>
+The Zephyr RTOS shall allow only the owning thread to unlock the mutex.
+<<<
```

Comment:
Avoid using the word "only" in requirements. It is usually(?) very hard or impossible to verify.

In many cI think this is better phrased as:

While the mutex is not in a recursive lock,
When the thread owning a mutex requests to unlock it,
the ... shall ...

Reply:
I think this is covered by my suggested wording for  ZEP-MUTEX-5  Unlocking of a Mutex:
"The Zephyr RTOS shall provide a mechanism for the owning thread to unlock a mutex."